pkg/kubevirt/actuator.go

```diff
@@ -0,0 +1,275 @@
+		msg := "unable to fetch root cloud cred secret"
+		logger.WithError(err).Error(msg)
+		return nil, &actuatoriface.ActuatorError{
+			ErrReason: minterv1.CredentialsProvisionFailure,
+			Message:   fmt.Sprintf("%v: %v", msg, err),
+		}
+	}
+
+	// get the secret data - the infra kubeconfig
+	infraClusterKubeconfig, ok := kubevirtCredentialsSecret.Data[KubevirtCredentialsSecretKey]
+	if !ok {
+		return nil, errors.New("invalid mode")
+	}
+
+	logger.Debug("extracted kubevirt credentials")
+	return infraClusterKubeconfig, nil
+}
+
+func (a *KubevirtActuator) syncCredentialSecret(ctx context.Context, cr *minterv1.CredentialsRequest, kubevirtCredentialData *[]byte, existingSecret *corev1.Secret, logger log.FieldLogger) error{
+	if existingSecret == nil {
+		if kubevirtCredentialData == nil {
+			msg := "new access key secret needed but no key data provided"
+			logger.Error(msg)
+			return &actuatoriface.ActuatorError{
+				ErrReason: minterv1.CredentialsProvisionFailure,
+				Message:   msg,
+			}
+		}
+
+		return a.createNewSecret(logger, cr, kubevirtCredentialData, ctx)
+	}
+
+	return a.updateExistingSecret(logger, existingSecret, cr, kubevirtCredentialData)
+}
+
+func (a *KubevirtActuator) updateExistingSecret(logger log.FieldLogger, existingSecret *corev1.Secret, cr *minterv1.CredentialsRequest, kubevirtCredentialData *[]byte) error {
+	// Update the existing secret:
+	logger.Debug("updating secret")
+	origSecret := existingSecret.DeepCopy()
+	if existingSecret.Annotations == nil {
+		existingSecret.Annotations = map[string]string{}
+	}
+	existingSecret.Annotations[minterv1.AnnotationCredentialsRequest] = fmt.Sprintf("%s/%s", cr.Namespace, cr.Name)
+	if kubevirtCredentialData != nil {
+		existingSecret.Data = map[string][]byte{
+			KubevirtCredentialsSecretKey: *kubevirtCredentialData,
+		}
+	}
+
+	if !reflect.DeepEqual(existingSecret, origSecret) {
+		logger.Info("target secret has changed, updating")
+		if err := a.Client.Update(context.TODO(), existingSecret); err != nil {
+			msg := "error updating secret"
+			logger.WithError(err).Error(msg)
+			return &actuatoriface.ActuatorError{
+				ErrReason: minterv1.CredentialsProvisionFailure,
+				Message:   msg,
+			}
+		}
+	} else {
+		logger.Debug("target secret unchanged")
+	}
+
+	return nil
+}
+
+func (a *KubevirtActuator) createNewSecret(logger log.FieldLogger, cr *minterv1.CredentialsRequest, kubevirtCredentialData *[]byte, ctx context.Context) error {
+	logger.Info("creating secret")
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Spec.SecretRef.Name,
+			Namespace: cr.Spec.SecretRef.Namespace,
+			Annotations: map[string]string{
+				minterv1.AnnotationCredentialsRequest: fmt.Sprintf("%s/%s", cr.Namespace, cr.Name),
+			},
+		},
+		Data: map[string][]byte{
+			KubevirtCredentialsSecretKey: *kubevirtCredentialData,
+		},
+	}
+
+	if err := a.Client.Create(ctx, secret); err != nil {
+		logger.WithError(err).Error("error creating secret")
+		return err
```

Comment:
probably should built up an ActuatorError here to return back to the credentialsrequest controller.

Reply:
Im building it as part of "sync" function that calling to "syncCredentialSecret" function that calling to that function "createNewSecret".

If returned an error from "syncCredentialSecret" function than building an ActuatorError (line 124)

Reply:
I see what you're saying.

I suppose what I see is that from syncCredentialSecret() it can return an ActuatorError directly, and updateExistingSecret() also returns an ActuatorError, but only createNewSecret() doesn't return an ActuatorError. So I have to spend time thinking about why this function is special.

I'm okay with all returns from syncCredentialSecret() being regular `error` or all returns being ActuatorError. Lets just be consistent.

Reply:
You're right, better being consistently.
I fixed it, thanks